app/models/voucher.rb

```diff
@@ -1,15 +1,136 @@
 # frozen_string_literal: false
 
 class Voucher < ApplicationRecord
+  acts_as_paranoid
+
+  include CalculatedAdjustments
+
   belongs_to :enterprise
 
+  has_many :adjustments,
+           as: :originator,
+           class_name: 'Spree::Adjustment',
+           inverse_of: :voucher,
+           dependent: :nullify
+
   validates :code, presence: true, uniqueness: { scope: :enterprise_id }
 
+  before_validation :add_calculator
+
+  def self.adjust!(order)
```

Comment:
This is a lot of calculation in a model. I agree that it's not the right fit for a calculator but it could go into a service object.